### ISSUE TITLE
inverse context creation always creates a language map

### ISSUE THREAD

**Issue by roptat** [w3c/json-ld-api#415]
In steps 3.9 to 3.17, the context creation algorithm sets the `type-map` and the `language-map`. However, because step 3.12 is in the middle, you will first have to execute one of the steps from 3.9 to 3.11 (or none if no condition matches) and then one of the steps form 3.13 to 3.17. When a term definition matches one of the above conditions, it will nevertheless have a `language-map` (most likely because of step 3.17) when it shouldn't.

In jsonld 1.0, the equivalent steps in the same algorithm are well separated, so I would suggest to move 3.12 before 3.9, and add "Otherwise, " at the beginning of step 3.13. This helps me pass a few tests, including t0006 and t0007.

**Comment by gkellogg** [comment]
I presume you're referring to the [Inverse Context Creation algorithm[https://w3c.github.io/json-ld-api/#algorithm-1].

It would be helpful to run through an example where this doesn't work properly.

Step 3.6 will create a _container map_ entry for _container_ if it doesn't exist initialized to something like the following:

```
{
  "container": {
    "@language": {},
    "@type": {},
    "@any": {"@none": "term"}
  }
}
```

Step 3.8 references the "@type" entry as _type map_ and either steps 3.10 or 3.12 references the "@language" entry as _language map_.

Subsequent steps add values to either the _type map_ or the _language map_, so all entries in the reverse context will have all three entries.

The inverse context has become an increasingly complex piece of code, due to the addition of `@direction` and greater use of containers; I would love to figure out how to simplify it in a future version.

**Comment by roptat** [comment]
Let me go through test 0006 then. Here is the data to compact:

```
{
  "@id": "http://example.org/id1",
  "@type": ["http://example.org/Type1", "http://example.org/Type2"],
  "http://example.org/term1": {"@value": "v1", "@type": "http://example.org/different-datatype"},
  "http://example.org/term2": {"@id": "http://example.org/id2"}
}
```

and here is the context to use:

```
{
  "@context": {
    "ex": "http://example.org/",
    "term1": { "@id": "ex:term1", "@type": "ex:datatype" },
    "term2": "ex:term2"
  }
}
```

During the inverse context creation, we will examine each key-value pair in the context, resulting in:

```
{
  "http://example.org/": {
    "@none": {
      "@language": {"@none": "ex"},
      "@type": {"@none": "ex"},
      "@any": {"@none": "ex"}
    }
  },
  "http://example.org/term1": {
    "@none": {
      "@language": {"@none": "term1"},
      "@type": {"http://example.org/datatype": "term1",
                "@none": "term1"},
      "@any": {"@none": "term1"},
    }
  },
  ...
}
```

Now, when compacting `http://example.org/term1`, I get `term1` instead of the expected `ex:term1` during the iri compaction of the term. During the term selection, the following parameters are passed by my implementation: `["@set", "@none", "@index", "@index@set"]` as `containers`, `"@type"` as type/language and `["http://example.org/different-datatype", "@none", "@any"]` as preferred values. With the above inverse context, `term1` is selected. With the proposed changes, I have the following inverse context:

```
{
  "http://example.org/": {
    "@none": {
      "@language": {"@none": "ex"},
      "@type": {"@none": "ex",
      "@any": {"@none" . "ex"}
    }
  },
  "http://example.org/term1": {
    "@none": {
      "@language": null,
      "@type": {"http://example.org/datatype": "term1"},
      "@any": {"@none" . "term1"}
    }
  },
  ...
}
```

Here, the difference is that `@type` contains only a mapping for `datatype`, not `@none` anymore, so `term1` cannot be selected. Instead, we use `https://example.org/` as the prefix, obtaining `ex:`.

I still have a lot of problems due to the iri compaction, or the inverse context, especially with containers, so one of these algorithms are surely wrong on my side. I'm not really sure what is expected, so tests and/or examples of the inverse context creation algorithm would be very welcome.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2020/2020-03-13-json-ld#section2-1).

- No actions or resolutions
<details><summary><i class="differentiate">View the transcript</i></summary>
2.1. [api] inverse context creation always creates a language map<br/>
Benjamin Young: <a href="https://github.com/w3c/json-ld-api/issues/415">https://github.com/w3c/json-ld-api/issues/415</a><br/>
Gregg Kellogg: possible issue on ordering things for inverse context creation<br/>
… will need to investigate before we can validate it<br/>
… some going back and forth on it<br/>
… asked for an example, and he cited compaction test 6 that exercises what he was getting at<br/>
… might need to reorder steps<br/>
… but could be a misunderstanding<br/>
</details>

**Comment by roptat** [comment]
Thank you. Indeed I'm not very confident this is an issue with the spec, it could be an implementation error, but it's really hard to understand where. The reason why I posted this, is because my proposed reordering of steps matches better jsonld 1.0, but there could be a reason why the two parts are separate, instead of conditional on one another.

Could someone share the correct inverse context in that example? It could help me determine where I got things wrong.

**Comment by gkellogg** [comment]
My implementation creates the following inverse context for compact/0006:

```json
{
  "http://example.org/": {
    "@none": {
      "@language": {"@none": "ex"},
      "@type": {"@none": "ex"},
      "@any": {"@none": "ex"}
    }
  },
  "http://example.org/term1": {
    "@none": {
      "@language": {},
      "@type": {"http://example.org/datatype": "term1"},
      "@any": {"@none": "term1"}
    }
  },
  "http://example.org/term2": {
    "@none": {
      "@language": {"@none": "term2"},
      "@type": {"@none": "term2"},
      "@any": {"@none": "term2"}
    }
  }
}
```
which seems to be consistent with what you come up with. I also use the same values for the other parameters, so the difference in interpretation seems to be in term selection rather than inverse context creation.

All maps contain only "@none", so that is the only container value which is examined and _value map_ will be `{"http://example.org/datatype":"term1"}`. As there is no entry for anything in _preferred values_ the _Term Selection_ algorithm returns no term, so the _IRI Compaction_ algorithm continues down to step 7 to end up using the "ex" prefix resulting in "ex:term1", which is the expected compact IRI.

**Comment by roptat** [comment]
so you obtain the same inverse context as I do *after* I change the inverse context creation algorithm as explained above. From that same inverse context, we obtain the same result in the term-selection phase, so that's good.

**Comment by gkellogg** [comment]
My implementation works pretty much the same as the spec, so you might want to [take a look at it](https://github.com/ruby-rdf/json-ld/blob/337af32e27e2629d17f8f7b87b14758704c954c3/lib/json/ld/context.rb#L1897-L1984)

**Comment by roptat** [comment]
in your implementation you have this `container = td.as_set? ? %(@set) : %(@none)` which seems to correspond to step 3.2. Why set it to `@set` though? it's not specified.

In the end, your implementation has a big `if`, `elsif` structure similar to what I propose:

`if td.reverse_property` (3.9)
`elsif td.type_mapping == '@none'` (3.10)
`elsif td.type_mapping` (3.11)
`elsif !td.language_mapping.nil? && !td.direction_mapping.nil?`(3.13)
`elsif !td.language_mapping.nil?` (3.14)
etc...

My reading of the specification lead me to think that step 3.13 should be an `if`, not an `elsif`, hence my suggestion to change it in the specification: move 3.12 so you can add "Otherwise" at the beginning of 3.13.

**Comment by gkellogg** [comment]
Okay, finally the light turns on 🤕 . 3.12 can move up to before 3.9 and 3.13 turn into an Otherwise if. I'll take care of that.

**Comment by gkellogg** [comment]
> in your implementation you have this `container = td.as_set? ? %(@set) : %(@none)` which seems to correspond to step 3.2. Why set it to `@set` though? it's not specified.

That is based on an early implementation decision to separate out `@set` from `@container` so that `@container` would be single-valued; something which didn't last, but remains in my implementation.

**Comment by gkellogg** [comment]
@roptat fixed in PR #423, let us know if that works.

**Comment by roptat** [comment]
it does, thank you!
